File: examples/go/table/saga-hand/splitter_example.go
// Saga splitter pattern example for documentation.
//
// Demonstrates the splitter pattern where one event triggers commands
// to multiple different aggregates.
package main

import (
	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"google.golang.org/protobuf/types/known/anypb"
)

// docs:start:saga_splitter
func handleTableSettled(event *examples.TableSettled, ctx *angzarr.SagaContext) ([]*pb.CommandBook, error) {
	// Split one event into commands for multiple player aggregates
	var commands []*pb.CommandBook

	for _, payout := range event.Payouts {
		cmd := &examples.TransferFunds{
			TableRoot: event.TableRoot,
			Amount:    payout.Amount,
		}

		targetSeq := ctx.GetSequence("player", payout.PlayerRoot)

		cmdAny, _ := anypb.New(cmd)
		commands = append(commands, &pb.CommandBook{
			Cover: &pb.Cover{Domain: "player", Root: &pb.UUID{Value: payout.PlayerRoot}},
			Pages: []*pb.CommandPage{{
				Sequence: &pb.CommandPage_Num{Num: targetSeq},
				Command:  cmdAny,
			}},
		})
	}

	return commands, nil // One CommandBook per player
}

// docs:end:saga_splitter

File: examples/cpp/table/saga-hand/src/splitter_example.cpp
// Saga splitter pattern example for documentation.
//
// Demonstrates the splitter pattern where one event triggers commands
// to multiple different aggregates.

#include <vector>

#include "angzarr/client.hpp"
#include "angzarr/proto/angzarr/types.pb.h"
#include "angzarr/proto/examples/table.pb.h"
#include "angzarr/proto/examples/player.pb.h"

using namespace angzarr;
using namespace angzarr::proto::angzarr;
using namespace angzarr::proto::examples;

// docs:start:saga_splitter
std::vector<CommandBook> handle_table_settled(
    const TableSettled& event, const SagaContext& ctx) {
    // Split one event into commands for multiple player aggregates
    std::vector<CommandBook> commands;

    for (const auto& payout : event.payouts()) {
        TransferFunds cmd;
        cmd.set_table_root(event.table_root());
        *cmd.mutable_amount() = payout.amount();

        uint32_t target_seq = ctx.get_sequence("player", payout.player_root());

        CommandBook book;
        auto* cover = book.mutable_cover();
        cover->set_domain("player");
        cover->mutable_root()->set_value(payout.player_root());

        auto* page = book.add_pages();
        page->set_num(target_seq);
        page->mutable_command()->PackFrom(cmd);

        commands.push_back(book);
    }

    return commands; // One CommandBook per player
}
// docs:end:saga_splitter
